packages/runtime/container-runtime/src/summaryStorageAdapter.ts

```diff
@@ -0,0 +1,171 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import { FetchSource, IDocumentStorageService, IDocumentStorageServicePolicies, ISummaryContext }
+    from "@fluidframework/driver-definitions";
+import { ICreateBlobResponse, ISnapshotTree, ISummaryBlob, ISummaryHandle,
+    ISummaryTree, IVersion, SummaryObject, SummaryType }
+    from "@fluidframework/protocol-definitions";
+/**
+ * This class uses Delegation Pattern to intercept methods of IDocumentStorageService implementation
+ * and adds the plugable postprocessing of summary data structures prior sending to Fluid Server
+ * and plugable preprocessing of summary data structures received from Fluid Server.
+ */
+class SummaryStorageAdapter implements IDocumentStorageService {
+    constructor(private readonly _delegate: IDocumentStorageService, private readonly _hooks: SummaryStorageHooks) { }
+    public get repositoryUrl(): string {
+        return this._delegate.repositoryUrl;
+    }
+    public get policies(): IDocumentStorageServicePolicies | undefined {
+        return this._delegate.policies;
+    }
+    public async getSnapshotTree(version?: IVersion | undefined, scenarioName?: string | undefined):
+        Promise<ISnapshotTree | null> {
+        const tree = await this._delegate.getSnapshotTree(version, scenarioName);
```

Comment:
It feels like code will be simpler if you do not use hooks, but rather provide noop adapter (implementing every method). And whoever needs to overwrite some methods simply overwrites them and calls base method (on base class).

Reply:
At the meeting "General Summary Compression Talk" on September 1st, We agreed that hooks are acceptable and we can benefit in the future from them (if additional pre / postprocessing is needed).

Reply:
@vladsud and I agree that intercepting IDocumentStorageService is a good pattern, but we would prefer you create a base adapter class that simply passes all calls through to the underlying service:

```ts
class SummaryStorageAdapter implements IDocumentStorageService {
    constructor (private readonly service) {}
    
    // Implement IDocumentStorageService by forwarding all calls to 'this.service'
    public async getSnapshotTree(...) { return this.service.getSnapshotTree(...); }
    ...
}
```

...and then implement then implement the CompressedStorageServiceAdapter by overriding 'getSnapshotTree'.

Multiple adapters are then explicitly chained, making the order in which they are applied clear:

```ts
new ChunkedStorageServiceAdapter(
    new CompressedStorageServiceAdapter(
        this.storage));
```

Reply:
> @vladsud and I agree that intercepting IDocumentStorageService is a good pattern, but we would prefer you create a base adapter class that simply passes all calls through to the underlying service:
> 
> ```ts
> class SummaryStorageAdapter implements IDocumentStorageService {
>     constructor (private readonly service) {}
>     
>     // Implement IDocumentStorageService by forwarding all calls to 'this.service'
>     public async getSnapshotTree(...) { return this.service.getSnapshotTree(...); }
>     ...
> }
> ```
> 
> ...and then implement then implement the CompressedStorageServiceAdapter by overriding 'getSnapshotTree'.
> 
> Multiple adapters are then explicitly chained, making the order in which they are applied clear:
> 
> ```ts
> new ChunkedStorageServiceAdapter(
>     new CompressedStorageServiceAdapter(
>         this.storage));
> ```

I like this one!

Reply:
Implemented in 86c95f5900c954f9897d37820080ea83e0f926fd. CompressedStorageServiceAdapter constructor has more parameters such as algorithm etc.